hack/generator/pkg/codegen/code_generator.go

```diff
@@ -20,33 +20,47 @@ type CodeGenerator struct {
 	pipeline      []PipelineStage
 }
 
-// NewCodeGeneratorFromConfigFile produces a new Generator with the given configuration file
-func NewCodeGeneratorFromConfigFile(configurationFile string) (*CodeGenerator, error) {
+// NewArmCodeGeneratorFromConfigFile produces a new Generator with the given configuration file
+func NewArmCodeGeneratorFromConfigFile(configurationFile string) (*CodeGenerator, error) {
 	configuration, err := config.LoadConfiguration(configurationFile)
 	if err != nil {
 		return nil, err
 	}
 
-	return NewCodeGeneratorFromConfig(configuration, astmodel.NewIdentifierFactory())
+	return NewArmCodeGeneratorFromConfig(configuration, astmodel.NewIdentifierFactory())
 }
 
-// NewCodeGeneratorFromConfig produces a new Generator with the given configuration
-func NewCodeGeneratorFromConfig(configuration *config.Configuration, idFactory astmodel.IdentifierFactory) (*CodeGenerator, error) {
-	var pipeline []PipelineStage
-	pipeline = append(pipeline, loadSchemaIntoTypes(idFactory, configuration, defaultSchemaLoader))
-	pipeline = append(pipeline, corePipelineStages(idFactory, configuration)...)
-	pipeline = append(pipeline, deleteGeneratedCode(configuration.OutputPath), exportPackages(configuration.OutputPath))
+// NewArmCodeGeneratorFromConfig produces a new Generator for Arm with the given configuration
+func NewArmCodeGeneratorFromConfig(configuration *config.Configuration, idFactory astmodel.IdentifierFactory) (*CodeGenerator, error) {
+	return NewCodeGeneratorFromConfig(
+		configuration,
+		idFactory,
+		[]PipelineStageTag{ImportStage, CoreStage, ArmStage, ExportStage, ReportingStage})
+}
+
+// NewCodeGeneratorFromConfig produces a new Generator for Arm with the given configuration
+func NewCodeGeneratorFromConfig(configuration *config.Configuration, idFactory astmodel.IdentifierFactory, includeTags []PipelineStageTag) (*CodeGenerator, error) {
+
+	allStages := createAllPipelineStages(idFactory, configuration)
 
 	result := &CodeGenerator{
 		configuration: configuration,
-		pipeline:      pipeline,
+	}
+
+	for _, s := range allStages {
+		if s.HasAnyTag(includeTags) {
+			result.pipeline = append(result.pipeline, s)
+		}
 	}
 
 	return result, nil
 }
 
-func corePipelineStages(idFactory astmodel.IdentifierFactory, configuration *config.Configuration) []PipelineStage {
+func createAllPipelineStages(idFactory astmodel.IdentifierFactory, configuration *config.Configuration) []PipelineStage {
 	return []PipelineStage{
+
+		loadSchemaIntoTypes(idFactory, configuration, defaultSchemaLoader),
```

Comment:
I think having the tags inside each stage is harder to understand/manage than if they were listed here. I realise the golden test files help with that (on the reading side at least), but my thinking is that the tags are organising information that belong at this level of the code, because this is the level that they're used. If we realised that we needed to exclude 5 core stages for a particular pipeline, it makes more sense to do that in one place than having to go and edit the stages to change their tags - the stages haven't changed at all, just the organising information used for assembling them has.

The tag of a stage isn't really inherent to the definition of the stage - when writing a new stage I wouldn't think "this is a core stage" or "this is a reporting stage", I'm just thinking about the specific thing that stage should do. From my perspective it's when I add them to the big list of stages that I'd want to make the decision about which group it should be in. (Although that's less clear for ArmStage or CrossplaneStage, I guess.)

Reply:
As discussed offline, have changed to using a `PipelineTarget` and putting the configuration into the creation of the pipeline itself, instead of distributing it across the stages themselves.